core/chains/evm/client/node_fsm.go

```diff
@@ -0,0 +1,224 @@
+package client
+
+import (
+	"context"
+	"fmt"
+)
+
+// NodeState represents the current state of the node
+// Node is a FSM
+type NodeState int
+
+func (n NodeState) String() string {
+	switch n {
+	case NodeStateUndialed:
+		return "Undialed"
+	case NodeStateDialed:
+		return "Dialed"
+	case NodeStateInvalidChainID:
+		return "InvalidChainID"
+	case NodeStateAlive:
+		return "Alive"
+	case NodeStateUnreachable:
+		return "Unreachable"
+	case NodeStateOutOfSync:
+		return "OutOfSync"
+	case NodeStateClosed:
+		return "Closed"
+	default:
+		return fmt.Sprintf("NodeState(%d)", n)
+	}
+}
+
+// GoString prints a prettier state
+func (n NodeState) GoString() string {
+	return fmt.Sprintf("NodeState%s(%d)", n.String(), n)
+}
+
+const (
+	// NodeStateUndialed is the first state of a virgin node
+	NodeStateUndialed = NodeState(iota)
+	// NodeStateDialed is after a node has successfully dialed but before it has verified the correct chain ID
+	NodeStateDialed
+	// NodeStateInvalidChainID is after chain ID verification failed
+	NodeStateInvalidChainID
+	// NodeStateAlive is a healthy node after chain ID verification succeeded
+	NodeStateAlive
+	// NodeStateUnreachable is a node that cannot be dialed or has disconnected
+	NodeStateUnreachable
+	// NodeStateOutOfSync is a node that is accepting connections but exceeded
+	// the failure threshold without sending any new heads. It will be
+	// disconnected, then put into a revive loop and re-awakened after redial
+	// if a new head arrives
+	NodeStateOutOfSync
+	// NodeStateClosed is after the connection has been closed and the node is at the end of its lifecycle
+	NodeStateClosed
+)
+
+// allNodeStates returns all possible states a node can be in
+// NOTE: Please update this when adding a new state above
+func allNodeStates() []NodeState {
```

Comment:
Is this just to range over? What about a single array/slice var  instead of allocating each time?

Reply:
Another option could be to define a `lenNodeState` after `NodeStateClosed` to use in regular for loops:
`for s:= NodeState(0);s<lenNodeState;s++`
Less to remember to maintain this way.

Reply:
Aha, now that works